Initial commit for Jakarta Security 4.0 - Multiple

daveywebster opened the pull request:
HttpAuthenticationMechanisms.

All functionality works for Jakarta Security 1/2/3.0 and 4.0.

A new service called HttpAuthenticationMechanismHandlerService was inserted into AuthModule for the interface methods validateRequest, cleanSubject and secureResponse.

The HAMHandlerService is fetched from the OSGi, and differs for Jakarta Security 1/2/3.0 and 4.0.

For Jakarta Security 1/2/3.0, the HAMHandlerService implements the identical functionality as previous for the methods methods above.

For Jakarta Security 4.0, the HAMHandlerService finds the implementation ( in-built or application specified ) of the
HttpAuthenticationmechansimHandler interface ( this is Jakarta 4.0 specified ) from the CDI, and defers to this implementation to execute the methods above.

For Jakarta Security 4.0, an additional completely internal HAM handler is created also with the @Alternative annotation to ensure the check for multiple HAM Handlers is skipped ( as per the specification ).

Fixes #32464

- [X] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [X] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).

daveywebster committed:
Initial commit for Jakarta Security 4.0 - Multiple
HttpAuthenticationMechanisms.

All functionality works for Jakarta Security 1/2/3.0 and 4.0.

A new service called HttpAuthenticationMechanismHandlerService was
inserted into AuthModule for the interface methods validateRequest,
cleanSubject and secureResponse.

The HAMHandlerService is fetched from the OSGi, and differs for Jakarta
Security 1/2/3.0 and 4.0.

For Jakarta Security 1/2/3.0, the HAMHandlerService implements the
identical functionality as previous for the methods methods above.

For Jakarta Security 4.0, the HAMHandlerService finds the implementation
( in-built or application specified ) of the
HttpAuthenticationmechansimHandler interface ( this is Jakarta 4.0
specified ) from the CDI, and defers to this implementation to execute
the methods above.

For Jakarta Security 4.0, an additional completely internal HAM handler
is created also with the @Alternative annotation to ensure the check for
multiple HAM Handlers is skipped ( as per the specification ).

daveywebster commented:
Here is a link to the design document (also mentioned in the Issue #32464 ).

[Multiple HttpAuthenticationMechanisms (HAMS) design document](https://ibm.box.com/s/tfse5nk8sft0szn4piw5t4frenxgscmh).

More generally, all developer related documentation to do with the Jakarta Security 4.0 implementation PRs and code can [be found here](https://ibm.ent.box.com/folder/320040894130).
TODO:

This PR is currently missing:

- Additional Junits
- Removal of two TODO items which point to the use of defined nlsprops messages other than free text
- HttpAuthenticationMechanisms new qualifiers attribute implementation.

NOTE: FAT test suite delivery is in another PR ( [documented here](https://github.com/OpenLiberty/open-liberty/issues/32649) ).

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a19ed2c0-1ca7-49e5-84ba-6cfd4d085a98) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

helyarp commented on dev/com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops:
```diff
@@ -29,6 +29,10 @@ JAVAEESEC_ERROR_NO_HAM=CWWKS1912E: The HttpAuthenticationMechanism object for th
 JAVAEESEC_ERROR_NO_HAM.explanation=The reason that the HttpAuthenticationMechanism object cannot be created varies. The error messages for the HttpAuthenticationMechanism object provide information on why the the HttpAuthenticationMechanism object cannot be created.
 JAVAEESEC_ERROR_NO_HAM.useraction=Investigate any error messages from the HttpAuthenticationMechanism object and make corrections based on the error messages.
 
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL=CWWKS1913E: The HttpAuthenticationMechanismService implementation object for the {0} module in the {1} application could not be found as an OSGi bundle.
```
Change "could not be found" to "was not found". The message would be clearer as "is not an OSGi bundle."

daveywebster replied:
Done, text as suggested.

helyarp commented on dev/com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops:
```diff
@@ -29,6 +29,10 @@ JAVAEESEC_ERROR_NO_HAM=CWWKS1912E: The HttpAuthenticationMechanism object for th
 JAVAEESEC_ERROR_NO_HAM.useraction=Investigate any error messages from the HttpAuthenticationMechanism object and make corrections based on the error messages.
 
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL=CWWKS1913E: The HttpAuthenticationMechanismService implementation object for the {0} module in the {1} application could not be found as an OSGi bundle.
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL.explanation=The reason that the HttpAuthenticationMechanismService implementation object cannot be found is that one of the dependent bundles containing this implementation must be missing from the features manifest file.
```
Suggested change: "The HttpAuthenticationMechanismService implementation object was not found because one of the dependent bundles that contains this implementation is missing from the features manifest file."

- Removes "The reason that"
- Replaces "containing" with "that contains" (gerunds get flagged by the quality checker)
- Replaces "must be" with "is"

daveywebster replied:
Done, changed as suggested.

helyarp commented on dev/com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops:
```diff
@@ -29,6 +29,10 @@ JAVAEESEC_ERROR_NO_HAM=CWWKS1912E: The HttpAuthenticationMechanism object for th
 
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL=CWWKS1913E: The HttpAuthenticationMechanismService implementation object for the {0} module in the {1} application could not be found as an OSGi bundle.
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL.explanation=The reason that the HttpAuthenticationMechanismService implementation object cannot be found is that one of the dependent bundles containing this implementation must be missing from the features manifest file.
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL.useraction=Ensure the feature manifest references a bundle which contains an implementation of the HttpAuthenticationMechanismService.
```
Change "which" to "that". 

Optionally, change "Ensure the" to "Ensure that the".

daveywebster replied:
Changed as suggested, including optionally.

helyarp reviewed (commented):
Suggested changes

daveywebster commented:
How I tested my changes (not including FATs which are part of a separate review):

Extensive JUnits have been completed for the 1.0/2.0/3.0 appSecurity bundles, and also the 4.0 bundle - these are in this PR.

**For Jakarta Security 1.0/2.0/3.0, I did the following:**

1. Create an application that uses a DatabaseIdentityStore, and uses <feature> (in server.xml), jakartaee-8.0 (appSecurity-3.0).
2. Built all changed OpenLiberty bundles and installed them locally.
3. In this application, I ensured all levels of  validation worked or failed as expected,  including invalid password, missing password, valid password, valid password but not authorized, and all passwords as plain text and encoded/encrypted.
4. I repeated the above for jakarta-9.1 and jakartaee-10.0 (appSecurity-4.0 and 5.0 respectively).
5. I tested missing OSGi HttpAuthenticationMechanismHandlerService (by removing it from the bundle manually).
6. For all manual tests, I checked the trace.log and server.log for unexpected error messages.

I then repeated steps 1-6 with an application that uses an LdapIdentityStore ( I have scripts to do all of this ).

**For Jakarta Security 4.0, I did the following:**

(note, the below applies to both a DatabaseIdentityStore and an LdapIdentityStore).

1. Create an application that uses a DatabaseIdentityStore, and uses <feature> (in server.xml), jakartaee-11.0 (appSecurity-4.0).  I specified the beta version in my pom.xml.
2. Built all changed OpenLiberty bundles and installed them locally.
3. In this application, I ensured all levels of  validation worked or failed as expected,  including invalid password, missing password, valid password, valid password but not authorized, and all passwords as plain text and encoded/encrypted. 
4. I tested 

- missing OSGi HttpAuthenticationMechanismHandlerService (by removing it from the bundle manually)
- no HAMs specified, and that the error message was from the nlsprops bundle
- ordering of multiple HAMs based on in-built HAMs as per documentation
- ordering of application HAMs based on priority
- ordering of all HAMs (ensuring application ones override in-built ones)
- missing CDI HttpAuthenticationMechanismHandler (by manual code trickery), and that the error message was from the nlsprops bundle
- ensure an application HttpAuthenticationMechanismHandler implementation overrode the in-built one
- missing HAMs, and that the error message was from the nlsprops bundle
- not able to determine a unique HAM, and that an AmbigiousResolutionException was thrown
- ensured the trace.log and messages.log showed a clear message regarding which HAM was deemed highest priority, and that the message was from an nlsprops bundle.

That's it!

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=fab8f675-1f47-4758-ade5-664a3848a014) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
Dear L2 Reviewer, 

Regarding changes within:

_dev/com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops_

- JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL has been reviewed previously, and I've made your suggested changes (although always happy for a re-review).

_dev/io.openliberty.security.jakartasec.4.0.internal/resources/io/openliberty/security/jakartasec/internal/resources/JakartaSecurity40Messages.nlsprops_

- This file is new for this PR, but exists in a previously reviewed PR: https://github.com/OpenLiberty/open-liberty/pull/32433 (which hasn't been committed to the code base yet due to outstanding code changes, but I needed to add new messages to this specific file it for this PR).

- JAKARTASEC_WARNING_PRODUCTION_USE is an existing entry that you already reviewed in the previous PR.
- JAKARTASEC_WARNING_ENV_VARIABLE_NOT_SET is an existing entry that you already reviewed in the previous PR.
- JAKARTASEC_ERROR_WRONG_CRED is an existing entry that you already reviewed in the previous PR.
- JAKARTASEC_WARNING_IDSTORE_CONFIG is an existing entry that you already reviewed in the previous PR.
- JAKARTASEC_ERROR_NO_HAM is copied directly from bundle com.ibm.ws.security.javaeesec (I have a similar situation to report on and needed to use my local bundle).
- JAKARTASEC_ERROR_AMBIGUOUS_RESOLUTION is new

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+        MultiHttpAuthenticationMechanism multiHttpAuthenticationMechanism = getHighestPriorityHttpAuthenticationMechanism();
+
+        if (multiHttpAuthenticationMechanism == null) {
+            if (tc.isDebugEnabled()) {
```
Should be a warning instead of debug given we have nothing to authenticate with

daveywebster replied:
Done

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+        }
+
+        if (multiHttpAuthenticationMechanisms.isEmpty()) {
+            if (tc.isDebugEnabled()) {
```
duplication of message - remove the debug statement

daveywebster replied:
Done

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+
+        // ... and compare their priorities
+        if (first.getPriority() == second.getPriority()) {
+            if (tc.isDebugEnabled()) {
```
Is this debug message adding anything over the error message given in this case we will always have the error message

daveywebster replied:
You're right, removed.

abutch3r commented on dev/com.ibm.ws.security.javaeesec/src/com/ibm/ws/security/javaeesec/AuthModule.java:
```diff
@@ -141,6 +139,27 @@ protected HttpMessageContext createHttpMessageContext(MessageInfo messageInfo, S
+     * @return The HttpAuthenticationMechanismHandlerService or null if one cannot be found.
+     */
+    protected HttpAuthenticationMechanismHandlerService getHttpAuthenticationMechanismHandlerService() {
+        if (tc.isDebugEnabled()) {
```
No Need for the debug statement here, the Entry/Exit trace which we add via the build will show us entering the method

daveywebster replied:
Okay, Done.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/src/io/openliberty/security/jakartasec40/cdi/extensions/JakartaSecurity40CDIExtension.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    public <T> void afterBeanDiscovery(@Observes AfterBeanDiscovery afterBeanDiscovery, BeanManager beanManager) {
+        if (tc.isDebugEnabled()) {
```
Automatically added entry/exit trace can handle this one given you aren't logging after in addition to the method name

daveywebster replied:
Yep, gotcha.  Done.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/resources/io/openliberty/security/jakartasec/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,45 @@
+# Message prefix block: 2600-2610
+
+JAKARTASEC_WARNING_PRODUCTION_USE=CWWKS2600W: An identity store that is configured by using an @InMemoryIdentityStoreDefinition annotation was detected within this application. Do not use this definition in a production environment.
+JAKARTASEC_WARNING_PRODUCTION_USE.explanation=Jakarta states to use @InMemoryIdentityStoreDefinition within the application for development, testing, and debugging purposes only.
```
Do not mention Jakarata here. just restate that the store definition is intended for non-production purposes

daveywebster replied:
Okay, Done.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/resources/io/openliberty/security/jakartasec/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,45 @@
+JAKARTASEC_ERROR_WRONG_CRED.useraction=Review the type of credential passed into the IdentityStore object by the HttpAuthMechanism implementation.
+
+#Note to translator, do not translate "EL"
+JAKARTASEC_WARNING_IDSTORE_CONFIG=CWWKS2603W: The Expression Language (EL) expression for the ''{0}'' attribute of the identity store annotation cannot be resolved to a valid value. Ensure that the EL expression and the result are valid and ensure that any referenced beans that are used in the expression are resolvable. The default attribute value of ''{1}'' is used instead.
```
2nd second is the useraction. I don't think it should be here, given that the admin looking at the logs will not be able to do anything in the case of someone causing this warning.

daveywebster replied:
Hi, not entirely sure what is to be done here.  I copied this message from existing messages file:

./com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops

where it is the same.

Can we remove <message>.useraction?  I thought we needed all three per <message>?

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/resources/io/openliberty/security/jakartasec/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,45 @@
+JAKARTASEC_ERROR_NO_HAM.explanation=The reason that the HttpAuthenticationMechanism object cannot be created varies. The error messages for the HttpAuthenticationMechanism object provide information on why the the HttpAuthenticationMechanism object cannot be created.
+JAKARTASEC_ERROR_NO_HAM.useraction=Investigate any error messages from the HttpAuthenticationMechanism object and make corrections based on the error messages.
+
+JAKARTASEC_ERROR_AMBIGUOUS_RESOLUTION=CWWKS2605E: Multiple HttpAuthenticationMechanisms specified, but the top two ( {0} and {1} ) both had the same priority of {2}.
```
`top two` will not translate well. as well as if I have 3 with the same priority, then you end up missing one off the list, so the message needs to accomodate a unknown number of HAMs with the same priority

Something like `Unable to determine which HttpAuthenticationMechanism to handle request. The following HttpAuthenticationMechanisms have the same priority of {0}: {1}` where {1} is the list of the HAMs.

abutch3r replied:
Also your explanation says other factors can have an impact, but the message focuses on one cause `priority` I assume if the other qualifiers were the cause we could only say which HAMs we couldn't decide between. so here, probably don't mention priority. but in the debug statements cover why we were unable to select a particular HAM.

daveywebster replied:
Have changed the message and also passing in a list of all HAMs which match the highest priority with a small  mention of the HAM type (which I suspect will come into play as I look to implement @Qualifiers).  See what you think.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+            return AuthenticationStatus.NOT_DONE;
+        }
+
+        Tr.info(tc, getApplicationName(), "Found the highest priority HttpAuthenticationMechanism: "
```
this will come out on every single request that comes through here. make this a debug message

daveywebster replied:
Yep, changed to whenever the HAM list is rebuilt (i.e. application restart, new HAM added/deleted/modified, etc ...).

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+
+        MultiHttpAuthenticationMechanism multiHttpAuthenticationMechanism = getHighestPriorityHttpAuthenticationMechanism();
+        if (multiHttpAuthenticationMechanism == null) {
+            if (tc.isDebugEnabled()) {
```
Should be a warning

daveywebster replied:
Okay, Done.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+
+        MultiHttpAuthenticationMechanism multiHttpAuthenticationMechanism = getHighestPriorityHttpAuthenticationMechanism();
+        if (multiHttpAuthenticationMechanism == null) {
+            if (tc.isDebugEnabled()) {
```
Should be a warning

daveywebster replied:
Yep, Done.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+        MultiHttpAuthenticationMechanism second = iterator.next();
+
+        // ... and compare their priorities
+        if (first.getPriority() == second.getPriority()) {
```
what happens here if I have more then two HAMs or the same priority, if i have multiple hams with the same priority I have to potentiall fix (Number of HAMs -1) times to get all of my HAMs to same place

daveywebster replied:
Yes, changed message and algorithm to check for X number of HAMs which match the top priority, not just the first two (which signals an error, but - as you mention - paints an incomplete picture).

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+                             + " both with priority " + first.getPriority());
+            }
+
+            Tr.error(tc, "JAKARTASEC_ERROR_AMBIGUOUS_RESOLUTION", first.getSimpleName(), second.getSimpleName(), first.getPriority());
```
In relation to my comments previously in the messages file for this error. Is this the only place we output this error? if it is, then we should focus the message on priority and not mention other potential factors, given we don't seem to check for anything within our handler. 

I assume other factors would be for customers own provided Multiple HAM handler where they will provide their own errors/exceptions/messages. Customers do not reuse our messages within their own code or user features

daveywebster replied:
As mentioned, have changed message with a small mention of HAM types which will come into play soon.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,385 @@
+        for (MultiHttpAuthenticationMechanism httpAuthenticationMechanism : multiHttpAuthenticationMechanisms) {
+            msg = msg.append(httpAuthenticationMechanism.getSimpleName()).append(" Priority = " + httpAuthenticationMechanism.getPriority()).append(prioritySeperator);
+            // remove trailing ", "
+            Tr.info(tc, getApplicationName(), msg.toString().substring(0, msg.length() - prioritySeperator.length()));
```
Should be debug. as I am unsure at this time how noisy this would be at info level and not something as a admin I would generally care about as it is more for the App Dev

daveywebster replied:
Yes, Done.  And was incorrectly in the `for` loop also.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/services/HttpAuthenticationMechanismHandlerServiceImpl.java:
```diff
@@ -0,0 +1,163 @@
+                                                                                     httpMessageContext.getResponse(),
+                                                                                     httpMessageContext);
+        } else {
+            if (tc.isDebugEnabled()) {
```
Is debug needed here given we will log the exception?

daveywebster replied:
Okay, removed.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/services/HttpAuthenticationMechanismHandlerServiceImpl.java:
```diff
@@ -0,0 +1,163 @@
+
+    @Override
+    public void cleanSubject(HttpMessageContext httpMessageContext) {
+        if (tc.isDebugEnabled()) {
```
Not needed

daveywebster replied:
Yep, removed.

abutch3r commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/services/HttpAuthenticationMechanismHandlerServiceImpl.java:
```diff
@@ -0,0 +1,163 @@
+                                                            httpMessageContext.getResponse(),
+                                                            httpMessageContext);
+        } else {
+            if (tc.isDebugEnabled()) {
```
would just remove the else, given the impact is that we don't do anything

daveywebster replied:
Okay, Done.

helyarp commented on dev/com.ibm.ws.security.javaeesec/resources/com/ibm/ws/security/javaeesec/internal/resources/JavaEESecMessages.nlsprops:
```diff
@@ -29,6 +29,10 @@ JAVAEESEC_ERROR_NO_HAM=CWWKS1912E: The HttpAuthenticationMechanism object for th
 JAVAEESEC_ERROR_NO_HAM.explanation=The reason that the HttpAuthenticationMechanism object cannot be created varies. The error messages for the HttpAuthenticationMechanism object provide information on why the the HttpAuthenticationMechanism object cannot be created.
 JAVAEESEC_ERROR_NO_HAM.useraction=Investigate any error messages from the HttpAuthenticationMechanism object and make corrections based on the error messages.
 
+JAVAEESEC_ERROR_NO_HAM_SERVICE_IMPL=CWWKS1913E: The HttpAuthenticationMechanismService implementation object for the {0} module in the {1} application was not be found as an OSGi bundle.
```
Change "was not be found" to "was not found" (remove be)

daveywebster replied:
Thanks and done.

helyarp reviewed (commented):
One small change requested

daveywebster commented:
Hi, still outstanding is:

- removal of OSGi for the HAMHandlerService
- add support for @Qualifiers

helyarp reviewed (commented):
Looks good!

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=273f410a-a945-4dcf-8d88-6f3a7ff32f2e) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
Note, still implementing the new `qualifiers` attribute of the `*AuthenticationMechanismDefinition` annotations.

tevans78 commented on dev/com.ibm.ws.security.javaeesec/test/com/ibm/ws/security/javaeesec/AuthModuleTest.java:
```diff
@@ -382,3 +392,5 @@ public HttpAuthenticationMechanism getHttpAuthenticationMechanism() {
 
 }
+
+// Made with Bob
```
Use of AI generated code is not currently permitted in Open Liberty.

daveywebster replied:
Ah, yep, bob helped with the tests, I modified them, forgot t remove the // Made ... 

cheers.

tevans78 replied:
Bob code was removed so resolving this conversation.

daveywebster commented:
Testing strategy :

Two separate applications exist:

1a/ Regression testing - application has an annotation defined, single Basic HAM and a user registry.
  * ensure authentication use cases (HTTP 200, 401, 403, exceptions) occur as expected during the auth work flow.
  * ensure exception raised for multiple HAMs (both annotation and application HAMs).
  * look at logging to see new log message ("HttpAuthenticationMechanismHandlerService 3 Using Jakarta Security 1.0/2.0/3.0 implementation")
 
1b/ Regression testing - same application and testing as above, but with a single Form HAM.
1c/ Regression testing - same appliation and testing as above, but using an OIDC HAM.
1d/ Regression testing - same application and testing as above, but using a custom HAM.

2/ New functionality - application has annotation defined HAMs: Basic, Form and Oidc (all in the same .java file).  Application also has three custom defined HAMs with different priorities.  Appliication also has a custom HAM Handler (with an algorithm which loads all available HAMs and chooses one at complete random) which is disabled initially.
* ensure authentication use cases  (HTTP 200, 401, 403, exceptions) occur as expected during the auth workflow.
  - manually confirm that the right custom HAM was used based on log messages from the custom HAMs.
* look at logging to see new log messages (one which outputs all HAMs in prioritised order and why the priority, another which outputs which HAM was choosen, and a third which outputs the name of the HAM Handler).

**_The following changes are done dynamically (i.e. the application wasn't explicitly stopped and restarted):_**

* now change the priority of the custom HAMs and ensure the highest priority is used.
* now remove the custom HAMs and ensure the Oidc HAM is used (as it is the highest available, check via log messages).
* now remove all the HAMs but one (leaving a single Basic or Form) and ensure the remaining one is used (again from log messages).
* re-add the custom HAMs, but have all three of them with the same priority and ensure the ambiguous ham exception is thrown with a message detailing the names of the three HAMs and their priorities).
* now enable the custom HAM handler and ensure it is selected for (via log output specific to the custom HAM handler).
---
Finally, go back to the original setup, run the authentication flows many times, and check the logs to ensure only a single message is output when appropriate.  For example, the output of the HAMs in prioritised order, the output of which HAM was selected, and the output of the HAM Handler should only ever be output when there is a change (or at the beginning).  Otherwise the logs will clog up.

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=77a21113-981f-43a7-afd6-ec8580979ca9) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
Testing Strategy:

In addition to the previous Testing Strategy comment, this latest PR contains qualifiers.  I manually tested qualifiers using the following:

- SCENARIO 1: Testing 3 BasicHAMs (two with qualifiers, one without), and a Custom HAM (with qualifier) and a CustomForm HAM (with qualifier) - and then, in the Custom HAM handler, ensuring all the HAMs are injected. NOTE: 4 qualifiers in total - Admin, User, Operator and Tester.
- SCENARIO 2: Testing 1 BasicHAM (without a qualifier) - and then, in the Custom HAM handler, ensuring it is injected using the @default annotation, and the same one can be selected without an annotation.
- SCENARIO 3: Testing 2 BasicHAMs (both with qualifiers) - and then, in the Custom HAM handler, ensuring the two HAMs are injected in with their qualifiers, and also one of them can be injected with the @default annotation and the same one can be selected without an annotation.
- Testing log file output to ensure a clear info message showing which HAMHandler was selected (it will be the name of the Custom HAM handler in the application).

---

**_SCENARIO1_**

In main application, define:

```
@BasicAuthenticationMechanismDefinition(realmName="admin-realm", qualifiers={Admin.class})
@BasicAuthenticationMechanismDefinition(realmName="user-realm", qualifiers={User.class})
@CustomFormAuthenticationMechanismDefinition(
	loginToContinue = @LoginToContinue(errorPage = "/operator-login-error.html",
		loginPage = "/operator-login.html"), qualifiers={Operator.class})
@FormAuthenticationMechanismDefinition(
	loginToContinue = @LoginToContinue(errorPage = "/login-error.html",
	loginPage = "/login.html"), qualifiers= {Tester.class})
@BasicAuthenticationMechanismDefinition(realmName="default-realm")
```

And in a custom handler, check that the Basic/Form/CustomForm HAMs were injected, and are different:

```
public class CustomHAMHandler implements HttpAuthenticationMechanismHandler {
	
	/**** SCENARIO 1 ****/

	@Inject @Admin
	private HttpAuthenticationMechanism adminHAM; 
	@Inject @User
	private HttpAuthenticationMechanism userHAM; 
 	@Inject @Operator
        private HttpAuthenticationMechanism operatorHAM;
	@Inject @Tester
	private HttpAuthenticationMechanism testerHAM;
	@Inject @Default
	private HttpAuthenticationMechanism defaultHAM;
	@Inject
	private HttpAuthenticationMechanism justHAM;

. . .

    // have to override validateRequest(), secureResponse() and cleanSubject() are optional to override.
    @Override
    public AuthenticationStatus validateRequest(HttpServletRequest request, 
                                                                                     HttpServletResponse response, 
                                                                                     HttpMessageContext httpMessageContext) {
    	
    	System.out.println("######## found adminHAM of [" + ((adminHAM == null) ? "null" : adminHAM.toString()) + "].");    	 
    	System.out.println("######## found userHAM of [" + ((userHAM == null) ? "null" : userHAM.toString()) + "].");    	
    	System.out.println("######## found operatorHAM of [" + ((operatorHAM == null) ? "null" : operatorHAM.toString()) + "].");    	    	
    	System.out.println("######## found testerHAM of [" + ((testerHAM == null) ? "null" : testerHAM.toString()) + "].");    	
    	System.out.println("######## found defaultHAM of [" + ((defaultHAM == null) ? "null" : defaultHAM.toString()) + "].");    	
    	System.out.println("######## found justHAM of [" + ((justHAM == null) ? "null" : justHAM.toString()) + "].");
```

**_To validate_**, I used a simple `curl` command which triggered the HAM injection and `validateRequest()` above.  In the above, `defaultHAM` and `justHAM` should have the same value, the rest should all be different.

---

**_SCENARIO2_**

I tested with just the single HAM:

```
@BasicAuthenticationMechanismDefinition(realmName="default-realm")
```

(and adjusted the HAM handler code appropriately)

**_To validate_**, I used a simple `curl` command which triggered the HAM injection and `validateRequest()` above.  In the above, defaultHAM and justHAM should have the same ID.  (NOTE: this is a slightly different code path, than any other set of HAM combinations, and also, tests the regression path).

---

**_SCENARIO3_**

I tested with HAMs: 

```
@BasicAuthenticationMechanismDefinition(realmName="admin-realm", qualifiers={Admin.class})
@BasicAuthenticationMechanismDefinition(realmName="user-realm", qualifiers={User.class})
```

(and adjusted the HAM handler code appropriately)

**_To validate_**, I used a simple `curl` command which triggered the HAM injection and `validateRequest()` above.  In the above, `defaultHAM` and `justHAM` should have the same value, the rest should all be different (NOTE: Again, this is a slightly different code path, picking a HAM to add the @Default to because core Liberty always needs a single default HAM).

**_SCENARIO3_**

I tested with HAMs: 

```
@OpenIdAuthenticationMechanismDefinition(providerURI = "https://provider1.com" ,qualifiers = {Admin.class})
@OpenIdAuthenticationMechanismDefinition(providerURI = "https://provider2.com", qualifiers = {User.class})
@OpenIdAuthenticationMechanismDefinition(providerURI = "https://provider3.com", qualifiers = {Operator.class})
@OpenIdAuthenticationMechanismDefinition(providerURI = "https://provider4.com" ,qualifiers = {Tester.class})
@OpenIdAuthenticationMechanismDefinition(providerURI = "https://provider5.com")
```

(and adjusted the HAM handler code appropriately)

**_To validate_**, I used a simple `curl` command which triggered the HAM injection and `validateRequest()` above.  In the above, `defaultHAM` and `justHAM` should have the same value, the rest should all be different (NOTE: Again, this is a slightly different code path, picking a HAM to add the @Default to because core Liberty always needs a single default HAM).

---

tevans78 commented on dev/com.ibm.ws.security.javaeesec.cdi/test/com/ibm/ws/security/javaeesec/cdi/extensions/JavaEESecCDIExtensionTest.java:
Please check all copyright dates.

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=63d5dd7c-7fa0-4cf0-aadb-14d486c84456) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
!build 
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=80d5fb66-b850-4f35-bbed-08cae8cd63ba) - :x: completed with errors/failures.)

(the previous build had 4 errors, but I can't replicate them locally and they **_appear_** to be unrelated to the code change, so am wondering if I needed to rebase and try again)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ddba22b2-5434-4b12-864b-2fa9ccfa2b84) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
Latest changes addresses the PB build break from earlier (there were three, two were already linked to defects, and the third one was defintely associated the the code changes in the PR).

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=768d417f-8ea5-4198-b683-689eeb835b48) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

**DW: Warning above relates to a single issue, which is linked to an ongoing PR in GHE.  No errors or failures.**

daveywebster commented:
Latest changes addresses code cleanup. 

Testing

---

### Manually regression tested all the single HAM cases.

**Test inputs**

In the main application, I had one HAM of type:

- Form
- CustomForm
- Basic
- Oidc
- Developer

**Expected results**

In all cases, the HAM was used as expected.

---

### Tested Multi-HAM, all different HAM types - no qualifiers.

**Test inputs**

In the main application, I had the following:

```
@BasicAuthenticationMechanismDefinition(realmName="basicAuth")

@FormAuthenticationMechanismDefinition(
loginToContinue = @LoginToContinue(errorPage = "/login-error.html",
loginPage = "/login.html"))


@OpenIdAuthenticationMechanismDefinition( providerURI = "https://samples.auth0.com/authorize",
clientId="kbyuFDidLLm280LIwVFiazOqjO3ty8KH",
redirectURI= "https://openidconnect.net/callback",
scope="openid profile email phone address",
responseType="code")
```
- Along with three developer created HAMs, each with **a different @Priority value**.

**Expected results**

- No error upon startup regarding multiple HAMs not allowed.
- Highest priority developer HAM chosen over all others.
- Clear log message regarding listing all the HAMs (6 in this case), how they were prioritized and which HAM was chosen.
- The log message from above was only shown only a single time per application (re)start, **not** after every auth work flow.
- Repeatedly removed a HAM until only one remaining to ensure a) the next highest one was chosen, and b) the log message was shown and updated (and again, only shown once when it changed).
- Removed the last HAM to ensure the code falls back to looking for a user registry.
- Added all the HAMs back into the code, but created two developer HAMs with the same @Priority value and ensured an error message was output regarding the inability to pick a single, prioritized HAM.  Also, removed all @Priority values from developer HAMs and checked if the same error was thrown (this is because if no @Priority value is given, a default one is assigned).

---

### Tested multiple HAMs with qualifiers.

**Test inputs**

In the main application, I had the following (sets only active one at a time):

```
/**** SET 1 ****/

@CustomFormAuthenticationMechanismDefinition(
	loginToContinue = @LoginToContinue(errorPage = "/operator-login-error.html",
		loginPage = "/operator-login.html"), qualifiers={Operator.class})
@FormAuthenticationMechanismDefinition(
	loginToContinue = @LoginToContinue(errorPage = "/login-error.html",
	loginPage = "/login.html"), qualifiers= {Tester.class})
@BasicAuthenticationMechanismDefinition(realmName="admin-realm", qualifiers={Admin.class})
@BasicAuthenticationMechanismDefinition(realmName="user-realm", qualifiers={User.class})

/**** END SET 1 ****/

/**** SET 2 ****/
//@BasicAuthenticationMechanismDefinition(realmName="default-realm")
/**** END SET 2 ****/

/**** SET 3 ****/
//@BasicAuthenticationMechanismDefinition(realmName="admin-realm", qualifiers={Admin.class})
//@BasicAuthenticationMechanismDefinition(realmName="user-realm", qualifiers={User.class})
//@BasicAuthenticationMechanismDefinition(realmName="default-realm")

/**** END SET 3 *****/

/**** SET 4 ****/

/***
@OpenIdAuthenticationMechanismDefinition(
	providerURI = "https://provider1.com" ,qualifiers = {Admin.class})
@OpenIdAuthenticationMechanismDefinition(
	providerURI = "https://provider2.com", qualifiers = {User.class})
@OpenIdAuthenticationMechanismDefinition(
	providerURI = "https://provider3.com", qualifiers = {Operator.class})
@OpenIdAuthenticationMechanismDefinition(
	providerURI = "https://provider4.com" ,qualifiers = {Tester.class})
***/

/**** END SET 4 ****/
```

And in a custom HAM handler, I had the following sets, only one active at a time

```
	
	/**** SET 1 ****/

	@Inject @Admin
	private HttpAuthenticationMechanism adminHAM; 
	@Inject @User
	private HttpAuthenticationMechanism userHAM; 
 	@Inject @Operator
        private HttpAuthenticationMechanism operatorHAM;
	@Inject @Tester
	private HttpAuthenticationMechanism testerHAM;

	/**** END SET 1 ****/
	
	/**** SET 2 ****/
/***
	@Inject @Default
	private HttpAuthenticationMechanism defaultHAM;
	@Inject
	private HttpAuthenticationMechanism justHAM;

	// to avoid compile errors
	private HttpAuthenticationMechanism adminHAM; 
	private HttpAuthenticationMechanism userHAM; 
	private HttpAuthenticationMechanism operatorHAM;
	private HttpAuthenticationMechanism testerHAM;
***/
    /**** END SET 2 ****/

	/**** SET 3 ****/
/***
	@Inject @Admin 
	private HttpAuthenticationMechanism adminHAM; 
	@Inject @User
	private HttpAuthenticationMechanism userHAM; 
	@Inject @Default
	private HttpAuthenticationMechanism defaultHAM;
	@Inject
	private HttpAuthenticationMechanism justHAM;

	// to avoid compile errors
	private HttpAuthenticationMechanism operatorHAM;
	private HttpAuthenticationMechanism testerHAM;
***/
	/**** END SET 3 ****/

	/**** SET 4 ****/

/***
	@Inject @Admin
	private HttpAuthenticationMechanism adminHAM; 
	@Inject @User
	private HttpAuthenticationMechanism userHAM; 
	@Inject @Operator
        private HttpAuthenticationMechanism operatorHAM;
	@Inject @Tester
	private HttpAuthenticationMechanism testerHAM;
***/
	/**** END SET 4 ****/
```

**Expected results**

- In the validateRequest() method, ensured all injected HAMs were different, and were of the correct type.
- Ihecked the annotation values (such as OIDC connection string) to ensure these were correctly associated with the different HAMs.
- Created a single HAM with two qualifiers and checked that - when injected into two different variables - they were the same object.
- Ensured CDI bean resolutions (ambiguity, can't find bean) still applied, for example, 
  - creating two HAMs with the same qualifier name (ambiguous) 
  - or injection without a qualified HAM.

---

Updated all FAT suite testing documents appropriately.

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6f136c21-2f36-4365-a2e3-0bd291d9623f) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

abutch3r commented on dev/io.openliberty.security.jakartasec.3.0.internal.cdi/src/io/openliberty/security/jakartasec/cdi/extensions/JakartaSecurity30CDIExtension.java:
```diff
@@ -74,20 +107,108 @@ protected void deactivate() {
+                addOidcHttpAuthenticationMechanismBean(oidcAnnotation, annotatedClass, annotatedType);
+            }
+        } catch (Exception e) {
+            if (tc.isDebugEnabled()) {
```
Should we be ignoring this exception if we actually get this issue? Also if this is an error, why would we only put the message out when debug is enabled? are there any particular exceptions that could occur here that we should be wary 

A few other examples where we only publish an exceptions information if debug is enabled, with no other form of handling.

daveywebster replied:
Am gonna re-work this new code into a 4.0 extension, and re-look at the error / debug output.

abutch3r commented on dev/io.openliberty.security.jakartasec.3.0.internal.cdi/src/io/openliberty/security/jakartasec/cdi/extensions/JakartaSecurity30CDIExtension.java:
```diff
@@ -1,10 +1,10 @@
 /*******************************************************************************
- * Copyright (c) 2022 IBM Corporation and others.
+ * Copyright (c) 2022, 2026 IBM Corporation and others.
```
Is there a reason why we are modifying the Jakarta Security 30 CDI extension, given we have the 40 CDI Extension?

abutch3r replied:
having multiple HAM code in the 30 extension looks like it would mean 30 now might support Mutliple HAMs, not 40

daveywebster replied:
Hi, yes, the Jakarta Security 30 CDI extension parses, handles OpenID (OIDC) HAMs.  Since we can now have a list of OIDC HAMs, the 30 extension was modified to parse the list and process each OIDC HAM.  An alternative was to create the 40 extension and do this separately, either duplicating the 30 extension code ("process each OIDC HAM") or create some common code (in the 30 bundle) that they both use.  This would be a less-risky approach, but it then duplicates code and introduces bug fixes in two places?  Or is there a better way I'm not considering - something CDI-ish or OSGi-ish?

abutch3r replied:
That is reasonable. will take another look through this class with that in mind

abutch3r reviewed (commented):
While I understand the need for an amount of code for Security 4.0 to be in common classes, there is a lot here and I am concerned we might have, in paritcular for the Jakarta30CDIExtensions put to much in other locations, rather then in 4.0 itself. 

If it can be explained why this needs to be in the base JavaEE code or Jakarta30 code then I can accept that.

utle commented on dev/com.ibm.ws.security.javaeesec/src/com/ibm/ws/security/javaeesec/AuthModule.java:
```diff
@@ -82,11 +85,16 @@ public void initialize(MessagePolicy requestPolicy, MessagePolicy responsePolicy
+            if (!JakartaSecurityValidationService.isJakartaSecurity40OrHigher()) {
+                authMech = getModulePropertiesUtils().getHttpAuthenticationMechanism();
+            }
+            AuthenticationStatus authenticationStatus = HttpAuthenticationMechanismHandlerService.validateRequest(httpMessageContext, authMech);
```
authMech can be null and it will cause NPE when calling the validateRequest

daveywebster replied:
Hi, yes I thought that , but here's the original code prior to any change:

```
 @Override
    public AuthStatus validateRequest(MessageInfo messageInfo, Subject clientSubject, Subject serviceSubject) throws AuthException {
        AuthStatus status = AuthStatus.SEND_FAILURE;
        try {
            HttpAuthenticationMechanism authMech = getModulePropertiesUtils().getHttpAuthenticationMechanism();
            HttpMessageContext httpMessageContext = createHttpMessageContext(messageInfo, clientSubject);
            AuthenticationStatus authenticationStatus = authMech.validateRequest(httpMessageContext.getRequest(),
                                                                                 httpMessageContext.getResponse(),
                                                                                 httpMessageContext);
        status = translateValidateRequestStatus(authenticationStatus);
        registerSession(httpMessageContext);
   . . . .

```
So `authMech` is fetched from the modulePropertiesUtils and used straight with the `validateRequest()` - now that doesn't necessarily mean it is correct, but the null pointer exception error case hasn't come up, as far as I know.

So in the "new" code, the HttpAuthenticationMechanismHandlerService basically does the same for pre JS 4.0, and the HttpAuthenticationMechanismHandlerService for JS 4.0 doesn't actually use `authMech` - the HAMs are selected as there can be multiple ones, and prioritized.  

But I do think a null check is good.  If it is null, then what should `authenticationStatus` be set to?  `AuthenticationStatus.SEND_FAILURE` or `AuthenticationStatus.CONTINUE` ?

jhanders34 replied:
Right, but now you are conditionally not getting authMech so it will be null for 4.0 and will give you the NPE that Ut is talking about.

daveywebster replied:
Right, let me take another look at the comment and null pointer issue, but right now, for 4.0, authMech - as pointed out - is not fetched, therefore it is null (for 4.0).  But ... in the 4.0 HttpAuthenticationMechanismServiceHandler code, authMech is not used at all - the HAMs are fetched fresh via CDI.select().  Jared, scroll up in our personal Slack chat to early November where I explained this piece of logic and pondered if there is another way around it, given the surrounding context.  So passing null to 4.0 has no effect as the variable is effectively ignored.  Ideally, I wouldn't pass `authMech` at all - for 4.0, CDI.select(), for < 4.0, then calculate `authMech` using ModuleProperties* classes - but this requires heavy lifting to move the ModuleProperties* into the bundle that AuthModule's 3.0 HAMServiceHandler bundle depends on which seemed excessive and a poor ROI as opposed to the above - but that was my query, is there a better way to do this, and can it be done in a re-factoring phase post beta?

utle commented on dev/com.ibm.ws.security.javaeesec/src/com/ibm/ws/security/javaeesec/AuthModule.java:
```diff
@@ -105,11 +113,16 @@ public AuthStatus secureResponse(MessageInfo messageInfo, Subject serviceSubject
+            if (!JakartaSecurityValidationService.isJakartaSecurity40OrHigher()) {
+                authMech = getModulePropertiesUtils().getHttpAuthenticationMechanism();
+            }
+            AuthenticationStatus authenticationStatus = HttpAuthenticationMechanismHandlerService.secureResponse(httpMessageContext, authMech);
```
The samething above .. authMech can be null and secureResponse does not check for null

daveywebster replied:
Yes, see comment above?

utle commented on dev/com.ibm.ws.security.javaeesec/src/com/ibm/ws/security/javaeesec/AuthModule.java:
```diff
@@ -123,9 +136,15 @@ public AuthStatus secureResponse(MessageInfo messageInfo, Subject serviceSubject
+        if (!JakartaSecurityValidationService.isJakartaSecurity40OrHigher()) {
+            authMech = getModulePropertiesUtils().getHttpAuthenticationMechanism();
+        }
+        HttpAuthenticationMechanismHandlerService.cleanSubject(httpMessageContext, authMech);
```
same above

daveywebster replied:
And above.

utle commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/identitystore/permissions/IdentityStorePermissionService.java:
```diff
@@ -21,9 +21,13 @@ public class IdentityStorePermissionService {
+    // static flag to ensure the debug message is logged only once
+    private static volatile boolean debugMessageLogged = false;
+
     public static void checkPermission(String name) {
```
checkPermission() ? this message just enable debugMessageLogged=true

daveywebster replied:
No, there's a Tr.debug within the conditional block.  The old code always output this message which is wrong - it should only be output once upon start (or restart).   Therefore, we initialize it to false and log only if we haven't output a message yet.

utle replied:
the method name should be something enableDebugMessageLog or something similar rather than checkPermission()?

jhanders34 replied:
It does check permission on pre Security 4.0.  In 4.0 there is no permission checking because security manager is removed.  So maybe some additional comments maybe would suffice.  The method name can't change because it has to match the exact same package, class name and method names as the pre 4.0 implementation of this class.

daveywebster replied:
Ah sorry Ut, I see what you are saying, and thanks Jared for clarifying.  I will add comments.

daveywebster commented:
> While I understand the need for an amount of code for Security 4.0 to be in common classes, there is a lot here and I am concerned we might have, in paritcular for the Jakarta30CDIExtensions put to much in other locations, rather then in 4.0 itself.
> 
> If it can be explained why this needs to be in the base JavaEE code or Jakarta30 code then I can accept that.

Yes, there is a better way I believe, leaving the pre-4.0 code alone (for the extensions), adding list parsing in new, 4.0 only code, and seeing what the commonality is afterwards, to decide if re-factoring is justified.  Let me look at this now.

daveywebster commented:
Hi, so latest push contains:

- reworked the qualfier extension processing, moving it in its entirety to the 4.0 CDI bundle, leaving the 3.0 internal and 3.0 CDI bundles virtually unchanged
- added code comments in IdentityStorePermissionService
- commented here in PR re: `authMech` being null - no associated code change
- better error handling in the reworked qualifier extension processing

Local testing and custom FATSuite executions all pass.  Will run a PB now.

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e49bf7ab-eba9-40e1-9e70-7117d0a6303a) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

utle commented on dev/com.ibm.ws.security.javaeesec.cdi/src/com/ibm/ws/security/javaeesec/cdi/beans/BasicHttpAuthenticationMechanism.java:
```diff
@@ -100,13 +113,19 @@ public AuthenticationStatus validateRequest(HttpServletRequest request,
             }
         }
+        if (props != null) {
+            realmName = (String) props.get(JavaEESecConstants.REALM_NAME);
```
I did not read the spec but just wonder why we only get/set the realm name for BasicHttpAuthenticationMechanism?  how do we create a subject later on without the realm name? do we have a default realm name?

abutch3r replied:
`Realm` is mentioned 4 times in the entire spec, solely in relation to `BasicHTTPAuthenicationMechanism` defaultRealm at part of [Section 4.2.1](https://jakarta.ee/specifications/security/4.0/jakarta-security-spec-4.0.pdf#%5B%7B%22num%22%3A97%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C631.65%2Cnull%5D) (Hopefully the link takes to the right section) Not sure what we do for the other types of HAM. but none of them mention a default realm

abutch3r reviewed (dismissed):
Thank you for the changes David. You have addressed all my reservations, as we have time between now and a target ga to address other key aspects like exceptions and some of the other messages. I am pleased to say I approve

daveywebster commented:
Hi All, thanks for looking through the PR and the insightful, useful comments!

Just waiting for the PB to finish - nothing has changed (regression wise) since the last two PBs which both finished with an unrelated single warning, but I still want to wait.  

Cheers again.

daveywebster commented:
Ah, build breaks due to the PR, glad I waited to see the results.  Will fix and update PR.

daveywebster commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=7b62fc2f-b3e2-4f9d-a27b-4c97e2758691) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

daveywebster commented:
Build break fixed, now re-running a PB to ensure all is okay.  A very obvious error though, sigh ....

helyarp commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/resources/io/openliberty/security/jakartasec/cdi/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+
+JAKARTA_SECURITY_CDI_ERROR_CUSTOM_QUALIFIERS_NO_HAM_HANDLER=CWWKS2610E: Missing a custom handler when using qualified HttpAuthenticationMechanisms.
+JAKARTA_SECURITY_CDI_ERROR_CUSTOM_QUALIFIERS_NO_HAM_HANDLER.explanation=If qualified HttpAuthenticationMechanisms are specified, then the default HttpAuthenticationMechanismHandler cannot determine which HAM to use, therefore a custom implementation of the HttpAuthenticationMechanismHandler must be provided.
+JAKARTA_SECURITY_CDI_ERROR_CUSTOM_QUALIFIERS_NO_HAM_HANDLER.useraction=Implement an HttpAuthenticationMechanismHandler using customized logic to select a unique HttpAuthenticationMechanism.
```
The "using" is not clear. Replace with "that uses".

daveywebster replied:
Addressed in [new PR here](https://github.com/OpenLiberty/open-liberty/pull/34029).

helyarp commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/resources/io/openliberty/security/jakartasec/cdi/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+JAKARTA_SECURITY_CDI_ERROR_CUSTOM_QUALIFIERS_NO_HAM_HANDLER.useraction=Implement an HttpAuthenticationMechanismHandler using customized logic to select a unique HttpAuthenticationMechanism.
+
+JAKARTASEC_CDI_ERROR_PROCESSING_HAM_LIST=CWWKS2611E: Processing annotations for the {0} definition within a list failed.
+JAKARTASEC_CDI_ERROR_PROCESSING_HAM_LIST.explanation=Whilst processing a list of HttpAuthenticationMechanism definitions, an error occured whilst extracting annotation values.
```
Change the first "whilst" to "while" and the second "whilst" to "when". Change "occured" to "occurred" (spelling).
> While processing a list of HttpAuthenticationMechanism definitions, an error occurred when extracting annotation values.

helyarp commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/resources/io/openliberty/security/jakartasec/cdi/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+JAKARTASEC_CDI_ERROR_PROCESSING_HAM_LIST.useraction=Investigate the cause of the exception and correct the application configuration.
+
+JAKARTASEC_CDI_ERROR_EXTRACTING_HAM_PROPERTIES=CWWKS2612E: Processing properties the {0} type failed.
+JAKARTASEC_CDI_ERROR_EXTRACTING_HAM_PROPERTIES.explanation=Whilst processing a specific HttpAuthenticationMechanism annotation, an error occured whilst extracting property values.
```
Same changes to whilst and occured as in a previous message. Change to:
> While processing a specific HttpAuthenticationMechanism annotation, an error occurred when extracting property values.

daveywebster replied:
Addressed in [new PR here](https://github.com/OpenLiberty/open-liberty/pull/34029).

helyarp commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/resources/io/openliberty/security/jakartasec/cdi/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+JAKARTASEC_CDI_ERROR_EXTRACTING_HAM_PROPERTIES.explanation=Whilst processing a specific HttpAuthenticationMechanism annotation, an error occured whilst extracting property values.
+JAKARTASEC_CDI_ERROR_EXTRACTING_HAM_PROPERTIES.useraction=Investigate the cause of the exception and correct the application configuration.
+
+JAKARTASEC_CDI_ERROR_CREATING_QUALIFIERS=CWWKS2613E: Could not create custom qualifier annotation {0} for qualifier class {1}.
```
Change "Could not" to "cannot" and use the `the variable noun` format, unless the variables are so long that it'd make the message difficult to read.
> Cannot create the {0} custom qualifier annotation  for the {1} qualifier class.

daveywebster replied:
Addressed in [new PR here](https://github.com/OpenLiberty/open-liberty/pull/34029).

helyarp commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/resources/io/openliberty/security/jakartasec/cdi/internal/resources/JakartaSecurity40Messages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+JAKARTASEC_CDI_ERROR_EXTRACTING_HAM_PROPERTIES.useraction=Investigate the cause of the exception and correct the application configuration.
+
+JAKARTASEC_CDI_ERROR_CREATING_QUALIFIERS=CWWKS2613E: Could not create custom qualifier annotation {0} for qualifier class {1}.
+JAKARTASEC_CDI_ERROR_CREATING_QUALIFIERS.explanation=Whilst creating a qualified HttpAuthenticationMechanism, an error occured whilst creating a custom qualifier annotation.
```
Same changes to whilst and occured as previously. Change to:
> While creating a qualified HttpAuthenticationMechanism, an error occurred when creating a custom qualifier annotation.

helyarp reviewed (commented):
Suggested changes

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,430 @@
+        MultiHttpAuthenticationMechanism multiHttpAuthenticationMechanism = getHighestPriorityHttpAuthenticationMechanism();
+        if (multiHttpAuthenticationMechanism == null) {
+            if (tc.isWarningEnabled()) {
+                Tr.warning(tc, "No HttpAuthenticationMechanism available.");
```
Warning messages are supposed to be translated.

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,430 @@
+        };
+
+        if (action != null) {
+            AccessController.doPrivileged(action);
```
SecurityManager is not supported for Security 4.0, so does this need to be in a doPriv block?

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,430 @@
+            }
+        };
+
+        return AccessController.doPrivileged(action);
```
Same doPriv question.

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,430 @@
+        if (tc.isDebugEnabled() == false) {
+            return;
+        }
+        StringBuffer msg = new StringBuffer("Order of HttpAuthenticationMechanisms found (the first one will be used if its prioritization is unique - @Priority for application HAMs and HAM type - Oidc/CustomForm/Form/Basic - for in-built HAMs): ");
```
You should use StringBuilder these days instead of StringBuffer.

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/handlers/HttpAuthenticationMechanismHandlerImpl.java:
```diff
@@ -0,0 +1,430 @@
+            msg = msg.append(prioritySeperator);
+        }
+        // remove trailing ", "
+        Tr.debug(tc, getApplicationName(), msg.toString().substring(0, msg.length() - prioritySeperator.length()));
```
Instead of doing a substring, you can trim the buffer with `msg.setLength(msg.length() - prioritySeparator.length())` before calling Tr.debug.

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal.cdi/bnd.bnd:
```diff
@@ -57,15 +58,34 @@ src: src, resources
     com.ibm.websphere.security;version=latest,\
     com.ibm.ws.security.token;version=latest
+
+-testpath: \
```
`.classpath` was updated to include the new test directory.

jhanders34 replied:
This is fixed in #34044

jhanders34 commented on dev/io.openliberty.security.jakartasec.4.0.internal/.classpath:
```diff
@@ -1,6 +1,7 @@
 <classpath>
-	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="src" output="bin" path="src"/>
+	<classpathentry kind="src" output="bin_test" path="test"/>
```
I missed that the `test` path was already in this .classpath.  It is fixed in PR #34044.
